### DISCUSSION TITLE
New feature requests for 2024

### DISCUSSION THREAD

**Discussion by helto4real** [net-daemon/netdaemon#1066] [Feature requests]
Hey guys, I wanna use this threads to discuss ideas for new features for NetDaemon. Could be great to have a list of potential new features that we can have as an idea bank if someone wants to contribute but not have clear ideas what they want to contribute to.

**Comment by geeForceOne** [comment]
Ability to create helpers through the API :)

**Reply by helto4real** [reply]
#1096 are one of more PRs to start implementing this features. As a start it will be added to the `Client` and then we have to figure out how the HassModel APIs should look like (if we decide to make them).

**Comment by C47hub** [comment]
I would like it if I could work with calendars better. 
Capturing the events of a calendar currently is hard and doesn't work very well for me.
Also I am unable to read the contents of a calendar.
Respond to a calendar event x time before or after the event etc like home assistant automations does allow we also cannot do.

**Reply by helto4real** [reply]
Can you provide details on what you tried to do and failed with please?

**Comment by vinnie1234** [comment]
WhenTurnOn and WhenTurnOff

`    public static IDisposable WhenTurnsOn<T, TAttributes>(
        this Entity<T, EntityState<TAttributes>, TAttributes> entity,
        Action<StateChange<T, EntityState<TAttributes>>> observer)
        where TAttributes : class
        where T : Entity<T, EntityState<TAttributes>, TAttributes>
        => entity.StateChanges().Where(c => c.Old?.IsOff() == true && (c.New?.IsOn() ?? false)).Subscribe(observer);`

`    public static IDisposable WhenTurnsOff<T, TAttributes>(
        this Entity<T, EntityState<TAttributes>, TAttributes> entity,
        Action<StateChange<T, EntityState<TAttributes>>> observer)
        where TAttributes : class
        where T : Entity<T, EntityState<TAttributes>, TAttributes>
        => entity.StateChanges().Where(c => c.Old?.IsOn() == true && (c.New?.IsOff() ?? false)).Subscribe(observer);`

**Reply by vinnie1234** [reply]
I just changed it a little bit for myself and add throttle :) 

`public static IDisposable WhenTurnsOn<T, TAttributes>(
        this Entity<T, EntityState<TAttributes>, TAttributes> entity,
        Action<StateChange<T, EntityState<TAttributes>>> observer, int throttleInSeconds = 0)
        where TAttributes : class
        where T : Entity<T, EntityState<TAttributes>, TAttributes>
        => entity.StateChanges().Throttle(TimeSpan.FromSeconds(throttleInSeconds)).Where(c => c.Old?.IsOff() == true && (c.New?.IsOn() ?? false))
            .Subscribe(observer);`

`  public static IDisposable WhenTurnsOff<T, TAttributes>(
        this Entity<T, EntityState<TAttributes>, TAttributes> entity,
        Action<StateChange<T, EntityState<TAttributes>>> observer, int throttleInSeconds = 0)
        where TAttributes : class
        where T : Entity<T, EntityState<TAttributes>, TAttributes>
        => entity.StateChanges().Throttle(TimeSpan.FromSeconds(throttleInSeconds)).Where(c => c.Old?.IsOn() == true && (c.New?.IsOff() ?? false))
            .Subscribe(observer);`

**Comment by iamandymcinnes** [comment]
Generic identifiers on entity types with similar functionality. i.e. light, switch, scene, automation, input_select have on and off properties, or you could look at the state, so those types plus binary_sensor have boolean states. Having some way to be able to mix types in config yaml properties out of the box would be nice. You might have entities that return a boolean as one collection, others that return a numeric value, or string etc.
```
     BooleanEntities:
        - light.lounge_floor_lamp
        - light.lounge_lamp
        - switch.lounge_tv
        - scene.watching_tv
 ```

**Comment by iamandymcinnes** [comment]
I'd love a nice out of the box logging implementation maybe something that gives you logbook integration by default. Taking it one step further in terms of the talk around having a usecase for an ND UI again, turning on/off automations is one thing or being able to conveniently focus on a particular automation might also be nice, but perhaps being able to see the logs related to a specific automation and maybe a history of when it was called, gives more of an argument for a UI. Additionally where you create entities via MQTT for an automation, it might be nice to see the associated entities for an automation or even the HA entities you are using in your automation i.e. list the entities and their states for each automation.

**Reply by helto4real** [reply]
The logging could be additional extension. Ether as part of the current logging extension or own. Yea agree this could be a valid use-case for a UI.

**Comment by J-HB** [comment]
I would suggest a nice way to create and interact with notifications especially the ones in android/ios.

**Reply by helto4real** [reply]
This is a great idea, do you have more details on what you want to see in such API?

**Comment by helto4real** [comment]
Thank you all for the suggestions so far.

**Comment by geeForceOne** [comment]
codegen to use CallServiceWithResponseAsync instead of CallService for the services that support response.

**Reply by helto4real** [reply]
This would require it be available in the service metadata. Not sure it is but that could be investigated ofc.

**Reply by helto4real** [reply]
Sadly it is not provides the structure of the response. The max we could do is is to return a JsonElement maybe if it has a response capability.

**Reply by helto4real** [reply]
@geeForceOne this is now implemented a few releases ago
